### DISCUSSION TITLE
WebLog is not descriptive of username/password problems

### DISCUSSION THREAD

**Discussion by pearson** [arendst/Tasmota#15420] [Ideas]
### PROBLEM DESCRIPTION
When a web password is set on a device, and another device attempts a `WebSend` (or similar) connection without a username/password, the receiving device's WebLog does not display an error.

In this case, WebLog (even at the highest logging level) will simply display `HTP: Command` with no other information. 

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): ESP12-F modules
- [X] Tasmota binary firmware version number used: 11.0.0, 11.1.0
  - [X] Pre-compiled
  - [ ] Self-compiled
- [X] Flashing tools used: `esptool.py`
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
22:02:52.989 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 26, Pld -99, Data 'Template; Module; GPIO 255'
22:02:53.057 SRC: Backlog
22:02:53.059 CMD: Grp 0, Cmd 'TEMPLATE', Idx 1, Len 0, Pld -99, Data ''
22:02:53.065 MQT: stat/tasmota_4EF16C/RESULT = {"NAME":"TL SS01S Swtch","GPIO":[0,0,0,0,288,576,0,0,224,32,0,0,0,0],"FLAG":0,"BASE":18}
22:02:53.272 SRC: Backlog
22:02:53.274 CMD: Grp 0, Cmd 'MODULE', Idx 1, Len 0, Pld -99, Data ''
22:02:53.279 MQT: stat/tasmota_4EF16C/RESULT = {"Module":{"0":"TL SS01S Swtch"}}
22:02:53.522 SRC: Backlog
22:02:53.524 CMD: Grp 0, Cmd 'GPIO', Idx 1, Len 3, Pld 255, Data '255'
22:02:53.533 MQT: stat/tasmota_4EF16C/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"288":"Led1"},"GPIO5":{"576":"LedLink_i"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"224":"Relay1"},"GPIO13":{"32":"Button1"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
22:03:39.735 CMD: Grp 0, Cmd 'STATUS', Idx 1, Len 1, Pld 0, Data '0'
22:03:39.741 MQT: stat/tasmota_4EFABC/STATUS = {"Status":{"Module":0,"DeviceName":"tasmota-switch-lower-staircase-bottom","FriendlyName":["Tasmota"],"Topic":"tasmota_4EFABC","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
22:03:39.776 MQT: stat/tasmota_4EFABC/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T02:46:47","StartupUTC":"2022-04-21T18:16:52","Sleep":50,"CfgHolder":4617,"BootCount":5,"BCResetTime":"2022-04-20T23:39:30","SaveCount":53,"SaveAddress":"F7000"}}
22:03:39.811 MQT: stat/tasmota_4EFABC/STATUS2 = {"StatusFWR":{"Version":"11.1.0(tasmota)","BuildDateTime":"2022-04-13T06:40:42","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"415/699"}}
22:03:39.831 MQT: stat/tasmota_4EFABC/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":3,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["IoT-Devices",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A190000000000","00000080","00006000","00004000"]}}
22:03:39.863 MQT: stat/tasmota_4EFABC/STATUS4 = {"StatusMEM":{"ProgramSize":626,"Free":376,"Heap":24,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"1640C8","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","04000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,56,62","Sensors":"1,2,3,4,5,6"}}
22:03:39.896 MQT: stat/tasmota_4EFABC/STATUS5 = {"StatusNET":{"Hostname":"tasmota-4EFABC-4460","IPAddress":"10.0.10.4","Gateway":"10.0.1.1","Subnetmask":"255.255.0.0","DNSServer1":"10.0.1.1","DNSServer2":"0.0.0.0","Mac":"60:01:94:AB:CD:EF","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
22:03:39.922 MQT: stat/tasmota_4EFABC/STATUS6 = {"StatusMQT":{"MqttHost":"10.0.1.2","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_4EFABC","MqttUser":"light1","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
22:03:39.945 MQT: stat/tasmota_4EFABC/STATUS7 = {"StatusTIM":{"UTC":"2022-04-21T21:03:39","Local":"2022-04-21T22:03:39","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"05:47","Sunset":"19:49"}}
22:03:39.966 MQT: stat/tasmota_4EFABC/STATUS10 = {"StatusSNS":{"Time":"2022-04-21T22:03:39"}}
22:03:39.974 MQT: stat/tasmota_4EFABC/STATUS11 = {"StatusSTS":{"Time":"2022-04-21T22:03:39","Uptime":"0T02:46:47","UptimeSec":10007,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":24,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"IoT-Devices","BSSId":"7E:8A:20:AB:CD:EF","Channel":1,"Mode":"11n","RSSI":90,"Signal":-55,"LinkCount":1,"Downtime":"0T00:00:04"}}}
22:03:42.482 MQT: tele/tasmota_4EFABC/STATE = {"Time":"2022-04-21T22:03:42","Uptime":"0T02:46:50","UptimeSec":10010,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"IoT-Devices","BSSId":"7E:8A:20:AB:CD:EF","Channel":1,"Mode":"11n","RSSI":88,"Signal":-56,"LinkCount":1,"Downtime":"0T00:00:04"}}

```
- [X] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
19:08:06.018 HTP: Command

```

### TO REPRODUCE
_Steps to reproduce the behavior:_

1. Set a web password on device A
2. Set weblog level to 4 or 5 on device A
3. Open Device A's console
4. Use `WebSend` on Device B to send a command to Device A **without** using a username and password.
5. Notice that Device A's console only shows `HTP: Command` and no other error notification.

### EXPECTED BEHAVIOUR
I would expect an error message to be logged that documents a failed username/password connection attempt. 

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
Having an error output would help those who are new to Tasmota commands such as `WebSend`, as the reason why `WebSend` is not working would be immediately apparent.

Additionally, the logging of failed connection attempts could have uses in security, especially if the logs are being sent to a syslog server.

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by arendst** [comment]
As the current WebSend implementation does not use ANY feedback it just doesn't know the other end isn't there.

What you are looking for is a full blown implementation with handshake et al. That is currently beyond the goal of websend simply due to code size increase.

**Reply by sfromis** [reply]
It could work without any feedback or handshake back to WebSend, if the target device (in this case A) produced a local message about the failure, like it does in case of the referer check being enabled to reject commands from another device (without the source device knowing what happened).

**Reply by pearson** [reply]
@sfromis has the right idea here, I think. The target/receiving device is already performing a username/password check. All that's required is a log entry when that check fails.

**Reply by arendst** [reply]
I'll dive in and see if I can make a logging. I expect to add it somewhere here:

```
  if (!WebAuthenticate()) {
    // Prefer authorization via HTTP header (Basic auth), if it fails, use legacy method via GET parameters
    char tmp1[33];
    WebGetArg(PSTR("user"), tmp1, sizeof(tmp1));
    char tmp2[strlen(SettingsText(SET_WEBPWD)) + 2];  // Need space for an entered password longer than set password
    WebGetArg(PSTR("password"), tmp2, sizeof(tmp2));

    if (!(!strcmp(tmp1, WEB_USERNAME) && !strcmp(tmp2, SettingsText(SET_WEBPWD)))) {
      WSContentBegin(401, CT_APP_JSON);
      WSContentSend_P(PSTR("{\"" D_RSLT_WARNING "\":\"" D_NEED_USER_AND_PASSWORD "\"}"));
      WSContentEnd();
      return;
    }
  }
```
BTW did you try with define USE_WEBSEND_RESPONSE enabled? This should have reported anything coming back from the device.

**Reply by arendst** [reply]
How about this message on the "other" device:

```
17:49:45.543 HTP: Command
17:49:45.550 HTP: Bad userid and/or password
```

Will release soon.

**Reply by arendst** [reply]
Released in development available for download in half an hour.

**Reply by pearson** [reply]
@arendst , I was using the pre-compiled binary. 

That message is perfect. Well done!
